packages/web/src/sideEffects/localFs/utils.js

```diff
@@ -1,18 +1,50 @@
+/*
+ * Compare the given files for equality; file path and file contents
+ * @return {Boolean} true if the files are the same
+ */
+const sameFile = (a, b) => {
+  if (a.fullPath === b.fullPath) {
+    if (typeof a.source === typeof b.source) {
+      if (a.source.byteLength) {
+        // compare ArrayBuffer contents
+        const aArray = new Uint8Array(a.source)
+        const bArray = new Uint8Array(b.source)
+        return aArray.length === bArray.length && aArray.every((v, i) => v === bArray[i])
+      } else {
+        // compare String contents
+        return a.length === b.length && a.source === b.source
+      }
+    }
+  }
+  return false
+}
+
 /**
- * returns list of files whose source has changed compared to the file with the same
- * name in memFs
- * @param {Object} reference - The reference hash of files, organized by name.
- * @param {Array} files - The list of files : need to have a source & name field.
- * @return array of files that changed
+ * Compare the contents of the reference and the new file lists.
+ * @see flattenFiles() below
+ * @param {Array} reference - the reference list of files
+ * @param {Array} files - the new list of files
+ * @return {Array} a list of the changed files
  */
-const changedFiles = (reference, files) => files.filter((file) => {
-  const matches = reference.filter((ref) => ref.fullPath === file.fullPath && ref.source !== file.source)
-  return matches.length > 0
-})
+const changedFiles = (reference, files) =>  {
+  const working = reference.slice()
+  return files.filter((file) => {
+    const found = working.findIndex((a, i) => sameFile(a, file))
+    if (found < 0) {
+      return true
+    }
+    working.splice(found, 1)
+    return false
+  })
+}
 
-const flattenFiles = (files) => {
+/*
+ * Create a flattened list of files from the given file heiarchy.
+ * @return {Array} a list of file entries
+ */
+const flattenFiles = (hierarchy) => {
   let result = []
-  files.forEach((fileOrFolder) => {
+  hierarchy.forEach((fileOrFolder) => {
     if (fileOrFolder.children) {
```

Comment:
Not super familiar with this.  Does it still work with empty folders?

Reply:
Good question... I try it but I don’t think an empty folder will initiate a ‘watch’ event.